### PR TITLE
Add Sybase connector

### LINKED ISSUE #2481: Add sybase connector

### PR DESCRIPTION
**Pull request by academy-codex**
fixes #2481 

Will be adding the supporting results and documentation in a couple days.

### CODE REVIEW

**Comment by ebyhr** [comment]
Thanks for your contribution! 
There is an exising PR https://github.com/prestosql/presto/pull/2976. Could you talk with @tooptoop4 ?

Also, I would suggest you join the community Slack https://prestosql.io/slack.html

File: presto-sybase/src/main/java/io.prestosql.plugin.sybase/SybaseClient.java [hunk 1]
```diff
@@ -0,0 +1,338 @@
...
+                    tableHandles.add(new JdbcTableHandle(
+                            schemaTableName,
+                            resultSet.getString("TABLE CAT"),
+                            resultSet.getString("TABLE SCHEM"),
+                            resultSet.getString("TABLE NAME")));
+                }
+                if (tableHandles.isEmpty()) {
+                    return null;
+                }
+                if (tableHandles.size() > 1) {
```
Comment (tooptoop4) [inline presto-sybase/src/main/java/io.prestosql.plugin.sybase/SybaseClient.java:264]: how could that happen? what does schemaTableName look like in the exception log?

File: presto-sybase/src/main/java/io.prestosql.plugin.sybase/SybaseClient.java [hunk 2]
```diff
@@ -0,0 +1,338 @@
...
+    }
+
+    private SybaseTableDescription loadTable(String tableDescriptionPath) {
+        Map<String, String> fileInfo = new HashMap<>();
+        try {
+            File file = new File(tableDescriptionPath);
+            BufferedReader br = new BufferedReader(new FileReader(file));
+            String st;
+            while ((st = br.readLine()) != null) {
+                String[] pairs = st.split(" *= *", 2);
```
Comment (tooptoop4) [inline presto-sybase/src/main/java/io.prestosql.plugin.sybase/SybaseClient.java:294]:
```while ((st = br.readLine()) != null) {
                 String[] pairs = st.split(" *= *", 2);
```

seems like a hack

File: presto-sybase/src/main/java/io.prestosql.plugin.sybase/SybaseClient.java [hunk 3]
```diff
@@ -0,0 +1,338 @@
...
+                connection.getCatalog(),
+                escapeNamePattern(schemaName, escape).orElse(null),
+                escapeNamePattern(tableName, escape).orElse(null),
+                new String[]{"TABLE", "VIEW", "MATERIALIZED VIEW", "FORGIEN TABLE"});
+    }
+
+    private SybaseTableDescription loadTable(String tableDescriptionPath) {
+        Map<String, String> fileInfo = new HashMap<>();
+        try {
+            File file = new File(tableDescriptionPath);
```
Comment (tooptoop4) [inline presto-sybase/src/main/java/io.prestosql.plugin.sybase/SybaseClient.java:290]: why does a file need to be written?

File: presto-sybase/src/main/java/io.prestosql.plugin.sybase/SybaseClient.java [hunk 4]
```diff
@@ -0,0 +1,338 @@
...
+            while ((st = br.readLine()) != null) {
+                String[] pairs = st.split(" *= *", 2);
+                fileInfo.put(pairs[0], pairs.length == 1 ? "" : pairs[1]);
+            }
+            return new SybaseTableDescription(
+                    fileInfo.get("schemaName"),
+                    fileInfo.get("tableName"),
+                    fileInfo.get("column"),
+                    !isNull(fileInfo.get("minValue")) ? NumberFormat.getInstance().parse(fileInfo.get("minValue")) : null,
+                    !isNull(fileInfo.get("maxValue")) ? NumberFormat.getInstance().parse(fileInfo.get("maxValue")) : null,
```
Comment (tooptoop4) [inline presto-sybase/src/main/java/io.prestosql.plugin.sybase/SybaseClient.java:302]: what do the min/max get used for?

File: presto-sybase/src/main/java/io.prestosql.plugin.sybase/SybaseClient.java [hunk 5]
```diff
@@ -0,0 +1,338 @@
...
+                    Integer.valueOf(fileInfo.get("numberPartitions"))
+            );
+
+        } catch (ParseException | IOException e) {
+            log.error("File path is invalid!!!");
```
Comment (tooptoop4) [inline presto-sybase/src/main/java/io.prestosql.plugin.sybase/SybaseClient.java:307]: log the file path too?

File: presto-sybase/src/main/java/io.prestosql.plugin.sybase/SybaseClient.java [hunk 6]
```diff
@@ -0,0 +1,338 @@
...
+            throw new PrestoException(JDBC_ERROR, e);
+        }
+    }
+
+    @Override
+    public void commitCreateTable(JdbcIdentity identity, JdbcOutputTableHandle handle) {
+    // SybaseIQ doesnot allow qualifying the target of a rename
```
Comment (tooptoop4) [inline presto-sybase/src/main/java/io.prestosql.plugin.sybase/SybaseClient.java:314]: typo does not

File: presto-sybase/src/main/java/io.prestosql.plugin.sybase/SybaseClient.java [hunk 7]
```diff
@@ -0,0 +1,338 @@
+package io.prestosql.plugin.sybase;
+
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import io.airlift.log.Logger;
+import io.prestosql.plugin.jdbc.*;
```
Comment (tooptoop4) [inline presto-sybase/src/main/java/io.prestosql.plugin.sybase/SybaseClient.java:10]: are the * imports necessary? not just here but in many places

File: presto-sybase/src/main/java/io.prestosql.plugin.sybase/SybaseTableDescription.java [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+
+    public String getTableName() {
+        return tableName;
+    }
+
+    public String getColumn() {
+        return column;
+    }
+
+    public Integer getNumberPartitions() {
```
Comment (tooptoop4) [inline presto-sybase/src/main/java/io.prestosql.plugin.sybase/SybaseTableDescription.java:48]: I haven't seen other connectors use this, how does it help?

File: presto-sybase/src/main/java/io.prestosql.plugin.sybase/SybaseClient.java [hunk 9]
```diff
@@ -0,0 +1,338 @@
...
+            .put(TINYINT, "tinyint")
+            .put(DOUBLE, "double precision")
+            .put(REAL, "real")
+            .put(VARBINARY, "varbinary")
+            .put(DATE, "date")
+            .put(TIME, "time")
+            .put(TIME_WITH_TIME_ZONE, "time with time zone")
+            .put(TIMESTAMP, "timestamp")
+            .put(VARCHAR, "varchar")
+            .build();
```
Comment (tooptoop4) [inline presto-sybase/src/main/java/io.prestosql.plugin.sybase/SybaseClient.java:67]: can other types map to unboundedvarchar? or even any of these fail to map then fallback to unboundedvarchar?

File: presto-sybase/src/main/java/io.prestosql.plugin.sybase/SybaseClient.java [hunk 10]
```diff
@@ -0,0 +1,338 @@
...
+                .append(" RENAME TO ")
+                .append(quoted(handle.getTableName()));
+
+        try (Connection connection = connectionFactory.openConnection(identity)) {
+            execute(connection, sql.toString());
+        } catch (SQLException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
```
Comment (tooptoop4) [inline presto-sybase/src/main/java/io.prestosql.plugin.sybase/SybaseClient.java:327]: does renaming just a  column on existing table work?

Comment (tooptoop4) [inline presto-sybase/src/main/java/io.prestosql.plugin.sybase/SybaseClient.java:327]: does create table (as) or copy empty table work?

File: presto-sybase/src/main/java/io.prestosql.plugin.sybase/SybaseClient.java [hunk 11]
```diff
@@ -0,0 +1,338 @@
...
+            .build();
+
+    @Inject
+    public SybaseClient(BaseJdbcConfig config, ConnectionFactory connectionFactory) {
+        super(config, "\"", connectionFactory);
+    }
+
+    @Override
+    public boolean isLimitGuaranteed() {
+        return true;
```
Comment (tooptoop4) [inline presto-sybase/src/main/java/io.prestosql.plugin.sybase/SybaseClient.java:76]: should SELECT TOP be there?

File: presto-sybase/src/main/java/io.prestosql.plugin.sybase/SybaseClient.java [hunk 12]
```diff
@@ -0,0 +1,338 @@
...
+            .put(SMALLINT, "smallint")
+            .put(TINYINT, "tinyint")
+            .put(DOUBLE, "double precision")
+            .put(REAL, "real")
+            .put(VARBINARY, "varbinary")
+            .put(DATE, "date")
+            .put(TIME, "time")
+            .put(TIME_WITH_TIME_ZONE, "time with time zone")
+            .put(TIMESTAMP, "timestamp")
+            .put(VARCHAR, "varchar")
```
Comment (tooptoop4) [inline presto-sybase/src/main/java/io.prestosql.plugin.sybase/SybaseClient.java:66]: does >30k char get mapped?

File: presto-sybase/src/main/java/io.prestosql.plugin.sybase/SybaseClient.java [hunk 13]
```diff
@@ -0,0 +1,338 @@
...
+    private static final Logger log = Logger.get(SybaseClient.class);
+    private static final Map<Type, String> SQL_TYPES = ImmutableMap.<Type, String>builder()
+            .put(BOOLEAN, "boolean")
+            .put(BIGINT, "bigint")
+            .put(INTEGER, "integer")
+            .put(SMALLINT, "smallint")
+            .put(TINYINT, "tinyint")
+            .put(DOUBLE, "double precision")
+            .put(REAL, "real")
+            .put(VARBINARY, "varbinary")
```
Comment (tooptoop4) [inline presto-sybase/src/main/java/io.prestosql.plugin.sybase/SybaseClient.java:61]: http://infocenter.sybase.com/help/index.jsp?topic=/com.sybase.infocenter.dc36271.1550/html/blocks/blocks42.htm mentions some more types

File: presto-sybase/src/main/java/io.prestosql.plugin.sybase/SybaseClient.java (hunk 10 above)
Comment (academy-codex) [reply presto-sybase/src/main/java/io.prestosql.plugin.sybase/SybaseClient.java:327]: yes

**Comment by tooptoop4** [comment]
@academy-codex would you be able to add aggregate pushdown like https://github.com/prestosql/presto/commit/193872a247c49099c6eeda4640f65e36222a4ad0#diff-e4b92558819e35928d8d20100000291dda1e15cf907c1d2f06ba4015b0afd6cc?

**Comment by academy-codex** [comment]
@tooptoop4 We are working on it. Have updated my branch with latest 341 changes for now.

**Comment by academy-codex** [comment]
> @academy-codex would you be able to add aggregate pushdown like [193872a#diff-e4b92558819e35928d8d20100000291dda1e15cf907c1d2f06ba4015b0afd6cc](https://github.com/prestosql/presto/commit/193872a247c49099c6eeda4640f65e36222a4ad0#diff-e4b92558819e35928d8d20100000291dda1e15cf907c1d2f06ba4015b0afd6cc)?

Hey @tooptoop4 ,
Have added aggregation pushdown for sybase.. working on numeric columns except for INT type. Need to explicitly handle that.

**Comment by academy-codex** [comment]
@findepi @ebyhr @martint Can we start conversation in this thread again. The jTDS driver works with Sybase and I have tested it as well with the aggregation pushdown. How do we go about merging this PR from here. Can you please guide me with the next items to complete to get this merged.

**Comment by findepi** [comment]
Ideally we should have some tests for a connector.
For a JDBC connector we typically have this

- smoke test (see TestPostgreSqlIntegrationSmokeTest)
- "distirbuted queries test" (see TestPostgreSqlDistributedQueries)
- data type mapping (see TestPostgreSqlTypeMapping)

for these, you will need a sybase instance. you can follow TestingPostgreSqlServer and PostgreSqlQueryRunner examples, perhaps using https://hub.docker.com/r/datagrip/sybase  with `org.testcontainers.containers.GenericContainer` .


To simplify review, i would recommend adding as little functionality and as much testing as possible.
For example, it's OK to defer aggregate pushdown implementation to another PR, and leave this one focused on solid connector base implementation.

**Comment by academy-codex** [comment]
> Ideally we should have some tests for a connector.
> For a JDBC connector we typically have this
> 
> * smoke test (see TestPostgreSqlIntegrationSmokeTest)
> * "distirbuted queries test" (see TestPostgreSqlDistributedQueries)
> * data type mapping (see TestPostgreSqlTypeMapping)
> 
> for these, you will need a sybase instance. you can follow TestingPostgreSqlServer and PostgreSqlQueryRunner examples, perhaps using https://hub.docker.com/r/datagrip/sybase with `org.testcontainers.containers.GenericContainer` .
> 
> To simplify review, i would recommend adding as little functionality and as much testing as possible.
> For example, it's OK to defer aggregate pushdown implementation to another PR, and leave this one focused on solid connector base implementation.

Hey I tried implementing the test cases with refactoring.
How do i use the docker thing ?

**Comment by ebyhr** [comment]
@academy-codex I suppose you can try `JdbcDatabaseContainer` as `TestingMemSqlServer`. Please let us know if it doesn't work as expected.

**Review by ebyhr — COMMENTED** [review]
> How do i specify the database, username and password ? 

Left comments about those information.

**Comment by academy-codex** [comment]
@ebyhr I want to get this rolling again. I have prted the code of sybase to trino and have tested against docker image.
Let me complete the test cases requirement and push a commit, then we can maybe start this thread again and close it ?

**Comment by ebyhr** [comment]
@academy-codex Yes, I would review once you rebase and push a new commit. There were some change around tests since last review, so you will need to take care of that (e.g. `BaseJdbcConnectorTest` was added).

**Comment by academy-codex** [comment]
Continued in PR #8488.

**Comment by findepi** [comment]
> Continued in PR #8488.

**Pull request closed without merging**